### PR TITLE
Add dice jaccard callbacks

### PR DESCRIPTION
**Pull request by ternaus**

### CODE REVIEW

File: catalyst/dl/callbacks/metrics/dice.py [hunk 1]
```diff
@@ -33,4 +38,71 @@ def __init__(
...
+class MulticlassDiceMetricCallback(Callback):
+    def __init__(self, prefix: str = "dice", input_key: str = "targets", output_key: str = "logits", **metric_params):
+        super().__init__(CallbackOrder.Metric)
+        self.prefix = prefix
+        self.input_key = input_key
+        self.output_key = output_key
+        self.metric_params = metric_params
+        self.confusion_matrix = None
+        self.class_names = metric_params["class_names"]  # dictionary {class_id: class_name}
+        self.class_prefix = metric_params["class_prefix"]
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/dice.py]: I think it's better to add this params to init one, like 
```
output_key: str = "logits",  class_names, class_prefix, **metric_params
```

**Comment by ternaus** [comment]
PTAL @Scitator

File: catalyst/dl/callbacks/metrics/iou.py [hunk 2]
```diff
@@ -41,4 +49,89 @@ def __init__(
...
+                f"{self.class_prefix}_{metric_name}"] = jaccard_value
+
+        state.metrics.epoch_values[state.loader_name]["mean"] = np.mean(
+            [x for x in batch_metrics.values()]
+        )
+
+        self._reset_stats()
+
+
+__all__ = ["IouCallback", "JaccardCallback", "MulticlassJaccardMetricCallback"]
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/iou.py]: could you please also make a symlink `MulticlassJaccardMetricCallback` -> `MulticlassIouMetricCallback `

File: catalyst/dl/callbacks/metrics/dice.py [hunk 3]
```diff
@@ -33,4 +43,86 @@ def __init__(
...
+
+    dice = (2 * true_positives + epsilon) / (
+        2 * true_positives + false_positives + false_negatives + epsilon
+    )
+
+    if not np.all(dice <= 1):
+        raise ValueError("Dice index should be less than 1")
+
+    if not np.all(dice >= 0):
+        raise ValueError("Dice index should be more than 1")
```
Comment (TezRomacH) [inline catalyst/dl/callbacks/metrics/dice.py:69]: more than 0

File: catalyst/dl/callbacks/metrics/dice.py [hunk 4]
```diff
@@ -33,4 +43,86 @@ def __init__(
...
+        class_prefix="",
+        **metric_params
+    ):
+        super().__init__(CallbackOrder.Metric)
+        self.prefix = prefix
+        self.input_key = input_key
+        self.output_key = output_key
+        self.metric_params = metric_params
+        self.confusion_matrix = None
+        self.class_names = class_names  # dictionary {class_id: class_name}
```
Comment (TezRomacH) [inline catalyst/dl/callbacks/metrics/dice.py:90]: can it be just list, where indices are class_ids?

File: catalyst/dl/callbacks/metrics/iou.py [hunk 5]
```diff
@@ -41,4 +49,89 @@ def __init__(
...
+
+    jaccard = (true_positives + epsilon) / (
+        true_positives + false_positives + false_negatives + epsilon
+    )
+
+    if not np.all(jaccard <= 1):
+        raise ValueError("Jaccard index should be less than 1")
+
+    if not np.all(jaccard >= 0):
+        raise ValueError("Jaccard index should be more than 1")
```
Comment (TezRomacH) [inline catalyst/dl/callbacks/metrics/iou.py:76]: 0

**Review by TezRomacH — CHANGES_REQUESTED** [review]

File: catalyst/dl/callbacks/metrics/dice.py (hunk 4 above)
Comment (ternaus) [reply catalyst/dl/callbacks/metrics/dice.py:90]: I can imagine cases, where I want some classes to be excluded.

File: catalyst/dl/callbacks/metrics/dice.py [hunk 6]
```diff
@@ -33,4 +43,86 @@ def __init__(
         )
 
 
-__all__ = ["DiceCallback"]
+def calculate_dice(tp_fp_fn_dict: dict) -> np.array:
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/dice.py]: could you please use 
```
def calculate_dice(true_positives: int?, false_positives: int?, false_negatives: int?)
```
to remove duplication like

```
true_positives = tp_fp_fn_dict["true_positives"]
false_positives = tp_fp_fn_dict["false_positives"]
false_negatives = tp_fp_fn_dict["false_negatives"]
```

File: catalyst/dl/callbacks/metrics/iou.py [hunk 7]
```diff
@@ -41,4 +49,89 @@ def __init__(
 
 JaccardCallback = IouCallback
 
-__all__ = ["IouCallback", "JaccardCallback"]
+
+def calculate_jaccard(tp_fp_fn_dict: dict) -> np.array:
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/iou.py]: looks like we can remove the code duplication here also

File: catalyst/dl/callbacks/metrics/iou.py [hunk 8]
```diff
@@ -41,4 +49,89 @@ def __init__(
...
+                f"{self.class_prefix}_{metric_name}"] = jaccard_value
+
+        state.metrics.epoch_values[state.loader_name]["mean"] = np.mean(
+            [x for x in batch_metrics.values()]
+        )
+
+        self._reset_stats()
+
+
+__all__ = ["IouCallback", "JaccardCallback", "MulticlassIOUMetricCallback"]
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/iou.py]: could you please also add jaracd alias like `MulticlassJaccardMetricCallback = MulticlassIOUMetricCallback `?

File: catalyst/dl/callbacks/metrics/iou.py [hunk 9]
```diff
@@ -41,4 +49,89 @@ def __init__(
...
+        if self.confusion_matrix is None:
+            self.confusion_matrix = confusion_matrix
+        else:
+            self.confusion_matrix += confusion_matrix
+
+    def on_loader_end(self, state: RunnerState):
+
+        tp_fp_fn_dict = calculate_tp_fp_fn(self.confusion_matrix)
+
+        batch_metrics: Dict = calculate_jaccard(tp_fp_fn_dict)
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/iou.py]: here you can use just `calculate_jaccard(**tp_fp_fn_dict)`

File: catalyst/dl/callbacks/metrics/dice.py [hunk 10]
```diff
@@ -33,4 +43,86 @@ def __init__(
...
+
+        if self.confusion_matrix is None:
+            self.confusion_matrix = confusion_matrix
+        else:
+            self.confusion_matrix += confusion_matrix
+
+    def on_loader_end(self, state: RunnerState):
+        tp_fp_fn_dict = calculate_tp_fp_fn(self.confusion_matrix)
+
+        batch_metrics: Dict = calculate_dice(tp_fp_fn_dict)
```
Comment (Scitator) [inline catalyst/dl/callbacks/metrics/dice.py]: so, you can use here `calculate_dice(**tp_fp_fn_dict)`

**Pull request merged**